src/NHibernate/Multi/QueryBatchItemBase.cs

```diff
@@ -149,61 +172,63 @@ public IEnumerable<ISqlCommand> GetCommands()
 
 						tmpResults.Add(o);
 					}
-					_loaderResults[index] = tmpResults;
-
-					if (index == _queryInfos.Count - 1)
-					{
-						InitializeEntitiesAndCollections(reader, hydratedObjects);
-					}
```

Comment:
Somewhat unrelated bug fix by the way here: this `InitializeEntitiesAndCollections(reader, hydratedObjects)` call would be missing if the last result happen to be cached (and not a previous result). But for handling cache put batching I had to move it as a distinct call from the query batcher, fixing this trouble here by the way.